libdrgn/program.c

```diff
@@ -207,12 +208,21 @@ drgn_program_check_initialized(struct drgn_program *prog)
 static struct drgn_error *has_kdump_signature(const char *path, int fd,
 					      bool *ret)
 {
-	char signature[KDUMP_SIG_LEN];
+	char signature[KDUMP_SIG_READ_LEN];
 	ssize_t r = pread_all(fd, signature, sizeof(signature), 0);
 	if (r < 0)
 		return drgn_error_create_os("pread", errno, path);
-	*ret = (r == sizeof(signature)
-		&& memcmp(signature, KDUMP_SIGNATURE, sizeof(signature)) == 0);
+	if (r != sizeof(signature)) {
+		*ret = false;
+		return NULL;
+	} else if (memcmp(signature, FLATTENED_SIGNATURE, FLATTENED_SIG_LEN) == 0) {
+		return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
+					 "The given file is in the makedumpfile flattened "
+					 "format, which Drgn does not support. Use "
+					 "'makedumpfile -R newfile <oldfile' to reassemble "
+					 "the file into a format Drgn can read.");
+	}
+	*ret = memcmp(signature, KDUMP_SIGNATURE, KDUMP_SIG_LEN) == 0;
```

Comment:
The conditions can be a little more concise and the intention slightly clearer like this:

```suggestion
	if (r >= FLATTENED_SIG_LEN
	    && memcmp(signature, FLATTENED_SIGNATURE, FLATTENED_SIG_LEN) == 0) {
		return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
					 "The given file is in the makedumpfile flattened "
					 "format, which Drgn does not support. Use "
					 "'makedumpfile -R newfile <oldfile' to reassemble "
					 "the file into a format Drgn can read.");
	}
	*ret = (r >= KDUMP_SIG_LEN
		&& memcmp(signature, KDUMP_SIGNATURE, KDUMP_SIG_LEN) == 0);
```